### PR TITLE
Expose netty's SSL Handshake Timeout - Fixes #2118

### LINKED ISSUE #2118: Feature request: Expose netty's SSL Handshake Timeout to the Vertx user
Hello,
This is a feature request; it will be nice to be able to set the ssl handshake timeout.

Thank you!

https://github.com/eclipse/vert.x/blob/38a19afe50917de3783bdd1512f2314b71cc1750/src/main/java/io/vertx/core/net/impl/NetClientImpl.java#L208

https://netty.io/4.0/api/io/netty/handler/ssl/SslHandler.html#setHandshakeTimeout-long-java.util.concurrent.TimeUnit-

### PR DESCRIPTION
**Pull request by BillyYccc**
This work allows you to configure SSL handshake timeout in NetClient/NetServer.

### CODE REVIEW

**Comment by BillyYccc** [comment]
In addition, I have no idea how this makes sense in SNI due to https://github.com/eclipse-vertx/vert.x/commit/df4a3dd3fdc32c384ef9c677cd191db22564f7f3 since Netty does not give a way to configure the timeout in SniHandler directly.

File: src/test/java/io/vertx/core/net/NetTest.java [hunk 1]
```diff
@@ -3333,6 +3361,102 @@ public void testServerNetSocketShouldBeClosedWhenTheClosedHandlerIsCalled() thro
...
+    client.close();
+
+    NetServerOptions serverOptions = new NetServerOptions().setSsl(true)
+      .setKeyStoreOptions(Cert.SERVER_JKS.get());
+    server = vertx.createNetServer(serverOptions);
+
+    NetClientOptions clientOptions = new NetClientOptions().setSsl(true)
+      .setTrustAll(true)
+      // set 1ms to produce timeout
+      .setSslHandshakeTimeout(1);
```
Comment (vietj) [inline src/test/java/io/vertx/core/net/NetTest.java]: I think we want a larger timeout and test with a regular NetServer (without SSL) that does nothing instead.

Comment (vietj) [reply src/test/java/io/vertx/core/net/NetTest.java]: that will provide a more reliable test

Comment (vietj) [reply src/test/java/io/vertx/core/net/NetTest.java]: by larger : 200ms is fine

File: src/test/java/io/vertx/core/net/NetTest.java [hunk 2]
```diff
@@ -3333,6 +3361,102 @@ public void testServerNetSocketShouldBeClosedWhenTheClosedHandlerIsCalled() thro
...
+        NetSocket socket = res.result();
+
+        assertFalse(socket.isSsl());
+        socket.upgradeToSsl(v -> {
+          // this handler will never be called because of failure of handshake
+          assertFalse(socket.isSsl());
+        });
+        // wait a bit to make sure the above handler will not be called
+        try {
+          Thread.sleep(2000);
```
Comment (vietj) [inline src/test/java/io/vertx/core/net/NetTest.java]: use setTimer isntead not blocking

**Comment by vietj** [comment]
comment : use for tests a regular TCP server with no SSL, that does nothing. that will simply block the TCP handshake on the client and produce the timeout

**Comment by BillyYccc** [comment]
Updated with https://github.com/eclipse-vertx/vert.x/pull/2714/commits/c20f1627d4c724678bdd986eff89a80446eb7c07.

**Comment by BillyYccc** [comment]
I've rebased to master to resolve the conflicts.

**Comment by vietj** [comment]
ok, I can see 1ms timeout, I think it should be larger to allow the client to connect to the server and get the connection established, i.e 50ms or 100ms would provide better tests

**Comment by BillyYccc** [comment]
Do you mean setting larger time out in server and assert the connection between server and client is successful established?

**Comment by jinqkim** [comment]
@vietj  
When will it be merged?

**Comment by vietj** [comment]
I overlooked this and I think for consistency purpose we need to have a `sslHandshakeTimeoutUnit` as like other time unit we have in the project.

also this should be rebased on master.

**Comment by BillyYccc** [comment]
@vietj I have updated the PR, can you please review?

File: src/test/java/io/vertx/core/net/NetTest.java [hunk 3]
```diff
@@ -3368,6 +3394,104 @@ public void testHalfCloseCallsEndHandlerAfterBuffersAreDelivered() throws Except
...
+        NetSocket socket = res.result();
+
+        assertFalse(socket.isSsl());
+        socket.upgradeToSsl(v -> {
+          // this handler will never be called because of failure of handshake
+          fail("this should never be called");
+          assertFalse(socket.isSsl());
+        });
+        // wait a bit to make sure the above handler will not be called
+        vertx.setTimer(2000, id -> {
```
Comment (vietj) [inline src/test/java/io/vertx/core/net/NetTest.java]: this test should not use a timer, instead it should set an exception handler on the NetSocket which will get called when the upgrade fails, this is faster and more reliable.

Comment (vietj) [reply src/test/java/io/vertx/core/net/NetTest.java]: have a look at NetTest#TLSTest

File: src/test/java/io/vertx/core/net/NetTest.java [hunk 4]
```diff
@@ -3368,6 +3394,104 @@ public void testHalfCloseCallsEndHandlerAfterBuffersAreDelivered() throws Except
...
+      .setSsl(true)
+      .setTrustAll(true)
+      .setSslHandshakeTimeout(200)
+      .setSslHandshakeTimeoutUnit(TimeUnit.MILLISECONDS);
+    client = vertx.createNetClient(clientOptions);
+
+    server.connectHandler(s -> {
+    }).listen(testAddress, ar -> {
+      assertTrue(ar.succeeded());
+      client.connect(testAddress, res -> {
```
Comment (vietj) [inline src/test/java/io/vertx/core/net/NetTest.java]: use `onFailure` construct for this test:

```
client.connect(testAddress, onFailure(err -> { ... });
```

this avoids to test the async result and gives directly the error cause

**Comment by BillyYccc** [comment]
@vietj updated with minors in https://github.com/eclipse-vertx/vert.x/pull/2714/commits/ad8cf84c376968816c39f57fdbcb82fae67ea838.

**Comment by BillyYccc** [comment]
closing as this has been ported to master with https://github.com/eclipse-vertx/vert.x/commit/5dd4de7dc877f58ef6ef61e54da0d527bee5923c.

**Pull request closed without merging**